### DISCUSSION TITLE
Confusion about custom_config.h

### DISCUSSION THREAD

**Discussion by vunhatchuong** [manna-harbour/miryoku#174] [Q&A] [answered]
My keyboard is Ferris Sweep, documents told me to copy the mapping macro definition to `custom_config.h`.

But idk where should I put the copy. Do I make a new folder in `layouts/community/split3x5_2`/ like drasha for split3x6_3? Or do I make it normally in `keyboard/ferris/keymaps/` or make a copy of `/users/manna-harbour_miryoku`.

**Comment by manna-harbour** [comment, marked as answer]
`custom_config.h` is https://github.com/manna-harbour/miryoku_qmk/blob/miryoku/users/manna-harbour_miryoku/custom_config.h. I'll add links in the docs.

What do you want to change?

Miryoku QMK support for `ferris/sweep` is described at https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#split_3x5_2, so the mapping macro in use is defined in https://github.com/manna-harbour/miryoku_qmk/blob/miryoku/layouts/community/split_3x5_2/manna-harbour_miryoku/config.h. To customise the mapping you'd copy the macro definition into https://github.com/manna-harbour/miryoku_qmk/blob/miryoku/users/manna-harbour_miryoku/custom_config.h, rename it, and edit as desired, as described at https://github.com/manna-harbour/miryoku/discussions/85. 

But the 3x5+2 mapping doesn't have any extra keys, so the only thing to change there is the relative position of the primary and secondary thumbs. That's not recommended if you also use boards with 3 thumb keys as the muscle memory will be incompatible.

**Reply by vunhatchuong** [reply]
I want to change a bit here and there from each llayer.

For example, I want to swap the `ESC` and `Tab` position on **Base**. Change the **Num** layer so it ascend rather than descend (7, 8, 9,...). And maybe more.

**Reply by manna-harbour** [reply]
In that case, see the Substitute Layer section in https://github.com/manna-harbour/miryoku/discussions/85. You'll edit `custom_config.h` but you won't need to customise the mapping macro. If you modify thumb keys also check if you need to update the thumb combos at https://github.com/manna-harbour/miryoku_qmk/blob/d92f5d177210891b34c9207a6aa19b900d50439c/users/manna-harbour_miryoku/manna-harbour_miryoku.c#L62-L91.

**Reply by vunhatchuong** [reply]
If I have to edit things outside `custom_config.h`, wouldn't it be better to make a copy of the` /users/` to avoid conflict in the future?

**Reply by vunhatchuong** [reply]
So final question, is this the correct format ? I remove Layer **Button** and all of its occurrence in my revamp but it still say missing definition. 

```C
#pragma once

#define LAYOUT_miryoku( \
K00,   K01,   K02,   K03,   K04,          K05,   K06,   K07,   K08,   K09, \
K10,   K11,   K12,   K13,   K14,          K15,   K16,   K17,   K18,   K19, \
K20,   K21,   K22,   K23,   K24,          K25,   K26,   K27,   K28,   K29, \
N30,   N31,   K32,   K33,   K34,          K35,   K36,   K37,   N38,   N39 \
) \
LAYOUT_split_3x5_2( \
K00,   K01,   K02,   K03,   K04,          K05,   K06,   K07,   K08,   K09, \
K10,   K11,   K12,   K13,   K14,          K15,   K16,   K17,   K18,   K19, \
K20,   K21,   K22,   K23,   K24,          K25,   K26,   K27,   K28,   K29, \
                     K33,   K34,          K35,   K36 \
)

#define MIRYOKU_LAYER_BASE \
KC_Q,              KC_W,              KC_F,              KC_P,              KC_B,              KC_J,              KC_L,              KC_U,              KC_Y,              KC_SCLN,           \
LGUI_T(KC_A),      LALT_T(KC_R),      LCTL_T(KC_S),      LSFT_T(KC_T),      KC_G,              KC_M,              LSFT_T(KC_N),      LCTL_T(KC_E),      LALT_T(KC_I),      LGUI_T(KC_O),      \
KC_Z,              ALGR_T(KC_X),      KC_C,              KC_D,              KC_V,              KC_K,              KC_H,              KC_COMM,           ALGR_T(KC_DOT),    KC_SLSH,           \
asdU_NP,              U_NP,              LT(U_MEDIA,KC_ESC),LT(U_NAV,KC_SPC),  LT(U_MOUSE,KC_TAB),LT(U_SYM,KC_BSPC),  LT(U_NUM,KC_ENT), LT(U_FUN,KC_DEL),  U_NP,              U_NP               \

#define MIRYOKU_LAYER_NUM \
KC_LBRC,           KC_1,              KC_2,              KC_3,              KC_RBRC,           U_NA,              TD(U_TD_U_BASE),   TD(U_TD_U_EXTRA),  TD(U_TD_U_TAP),    TD(U_TD_BOOT),     \
KC_QUOT,           KC_4,              KC_5,              KC_6,              KC_EQL,            U_NA,              KC_LSFT,           KC_LCTL,           KC_LALT,           KC_LGUI,           \
KC_GRV,            KC_7,              KC_8,              KC_9,              KC_BSLS,           U_NA,              TD(U_TD_U_NUM),    TD(U_TD_U_NAV),    KC_ALGR,           U_NA,              \
U_NP,              U_NP,              KC_DOT,            KC_0,              KC_MINS,           U_NA,              U_NA,              U_NA,              U_NP,              U_NP

#define MIRYOKU_LAYER_SYM \
KC_LBRC,           KC_EXLM,           KC_AT,           KC_HASH,           KC_RBRC,           U_NA,              TD(U_TD_U_BASE),   TD(U_TD_U_EXTRA),  TD(U_TD_U_TAP),    TD(U_TD_BOOT),     \
KC_QUOT,           KC_DLR,            KC_PERC,           KC_CIRC,           KC_EQL,           U_NA,              KC_LSFT,           KC_LCTL,           KC_LALT,           KC_LGUI,           \
KC_GRV,            KC_AMPR,           KC_ASTR,             KC_LPRN,           KC_BSLS,           U_NA,              TD(U_TD_U_SYM),    TD(U_TD_U_MOUSE),  KC_ALGR,           U_NA,              \
U_NP,              U_NP,              KC_LPRN,           KC_RPRN,           KC_MINS,           U_NA,              U_NA,              U_NA,              U_NP,              U_NP

#define MIRYOKU_FUN \
KC_F10,            KC_F1,             KC_F2,             KC_F3,             KC_PSCR,           U_NA,              TD(U_TD_U_BASE),   TD(U_TD_U_EXTRA),  TD(U_TD_U_TAP),    TD(U_TD_BOOT),     \
KC_F11,            KC_F4,             KC_F5,             KC_F6,             KC_SCRL,           U_NA,              KC_LSFT,           KC_LCTL,           KC_LALT,           KC_LGUI,           \
KC_F12,            KC_F7,             KC_F8,             KC_F9,             KC_PAUS,           U_NA,              TD(U_TD_U_FUN),    TD(U_TD_U_MEDIA),  KC_ALGR,           U_NA,              \
U_NP,              U_NP,              KC_APP,            KC_SPC,            KC_TAB,            U_NA,              U_NA,              U_NA,              U_NP,              U_NP

#define MIRYOKU_NAV \
TD(U_TD_BOOT),     TD(U_TD_U_TAP),    TD(U_TD_U_EXTRA),  TD(U_TD_U_BASE),   U_NA,              U_RDO,             U_PST,             U_CPY,             U_CUT,             U_UND,             \
KC_LGUI,           KC_LALT,           KC_LCTL,           KC_LSFT,           U_NA,              CW_TOGG,           KC_LEFT,           KC_DOWN,           KC_UP,             KC_RGHT,           \
U_NA,              KC_ALGR,           TD(U_TD_U_NUM),    TD(U_TD_U_NAV),    U_NA,              KC_INS,            KC_HOME,           KC_PGDN,           KC_PGUP,           KC_END,            \
U_NP,              U_NP,              U_NA,              U_NA,              U_NA,              KC_BSPC,            KC_ENT,           KC_DEL,            U_NP,              U_NP

#define MIRYOKU_MOUSE \
TD(U_TD_BOOT),     TD(U_TD_U_TAP),    TD(U_TD_U_EXTRA),  TD(U_TD_U_BASE),   U_NA,              U_RDO,             U_PST,             U_CPY,             U_CUT,             U_UND,             \
KC_LGUI,           KC_LALT,           KC_LCTL,           KC_LSFT,           U_NA,              U_NU,              KC_MS_L,           KC_MS_D,           KC_MS_U,           KC_MS_R,           \
U_NA,              KC_ALGR,           TD(U_TD_U_SYM),    TD(U_TD_U_MOUSE),  U_NA,              U_NU,              KC_WH_L,           KC_WH_D,           KC_WH_U,           KC_WH_R,           \
U_NP,              U_NP,              U_NA,              U_NA,              U_NA,              KC_BTN1,           KC_BTN2,           KC_BTN3,           U_NP,              U_NP

#define MIRYOKU_LAYER_LIST \
MIRYOKU_X(BASE,   "Base") \
MIRYOKU_X(EXTRA,  "Extra") \
MIRYOKU_X(TAP,    "Tap") \
MIRYOKU_X(NAV,    "Nav") \
MIRYOKU_X(MOUSE,  "Mouse") \
MIRYOKU_X(MEDIA,  "Media") \
MIRYOKU_X(NUM,    "Num") \
MIRYOKU_X(SYM,    "Sym") \
MIRYOKU_X(FUN,    "Fun")
```

**Reply by manna-harbour** [reply]
> If I have to edit things outside `custom_config.h`, wouldn't it be better to make a copy of the` /users/` to avoid conflict in the future?

At the moment, combos have been added manually. At some point they'll be generated automatically, and also be overridable in `custom_config.h`. See https://github.com/manna-harbour/miryoku/issues/94 and https://github.com/manna-harbour/miryoku/issues/56.

With `custom_config.h`, you still need to manage updates, as keycodes might change, you just avoid merge conflicts. You can copy everything to new directories and rename, but that'll make it even more difficult to manage updates. The best place for modifying layers is actually in Miryoku Babel, as the layer tables don't change so there won't be merge conflicts.

**Reply by manna-harbour** [reply]
> So final question, is this the correct format ? I remove Layer **Button** and all of its occurrence in my revamp but it still say missing definition.

If you haven't modified the mapping you don't need to include it here. If you have, it needs to be renamed to `MIRYOKU_MAPPING`.

`MIRYOKU_FUN` should be `MIRYOKU_LAYER_FUN`, etc.

Button is also referenced from Extra, so you'd also need to substitute or remove Extra, or alias Button to another layer.

If you still have errors, push your fork and run a workflow build, and I'll be able to see the errors and exactly what you've changed.

**Reply by vunhatchuong** [reply]
Yeah `EXTRA` is the problem here, remove that fix it :).
